### DISCUSSION TITLE
limactl start vs create - where to draw the line?

### DISCUSSION THREAD

**Discussion by jay7x** [lima-vm/lima#1647] [Q&A] [answered]
As you may see in the latest commits `limactl create` was introduced recently in HEAD (not released yet atm). This commands mostly just creates the VM definition currently. Then `limactl start` do the rest (download files, do some sanity checks, run the VM).

Question is.. where to draw the line between those commands.

Discussion is transferred from [slack](https://cloud-native.slack.com/archives/C043N6ZFV9S/p1688066890156059):

> Yury Bushmelev
> though, as I see, all the downloads happen in the start action anyway
> 
> Jan Dubois
> That is an interesting point. I noticed some of our parameter validation also only happens when you start (e.g. checking that you cannot use virtiofs with qemu on macOS). I wonder if all the downloading and checking should happen at `create` time
> 
> Jan Dubois
> You can run `create`, `edit`, `validate` and still have the instance fail to start up because of configuration errors. That feels wrong.
> 
> Yury Bushmelev
> and maybe I’d like to have ability to --download-only :slightly_smiling_face:
> 
> Jan Dubois
> idk, I would like to see this discussed in a Github issue: should `create` download images right away? Should `edit` be able to change the image (sounds dangerous)? I don't have a settled opinion on these things, but they feel like they would fit into a more composable "API" like we get by separating `create` from `start`. Creating the disk image logically belongs to `create` in my mind.
> 
> Yury Bushmelev
> I have no strong opinion here.. I see the point to have all the things under `start` but it’s ok to expect to have them under `create` too.. let’s have an issue or discussion for this
> 
> Jan Dubois
> E.g. once you have created the instance, you should be able to start it even when you are offline.

Please share your thoughts! Thank you!

**Comment by AkihiroSuda** [comment]
👍  on let `limactl create` download the images and do some sanity checks. i.e., everything except running the VM

(PR is welcome)

**Reply by afbjorklund** [reply]
I also think that create should create the disk image (e.g. qemu-img), and that start should start the virtual machine (e.g. qemu-system). My only point was that there would still be users that would like the "create" to also start it, using a special flag. And that some users will continue to use "start" on a non-existing machine, and would appreciate it being created or get a question about it.

**Reply by jandubois** [reply]
> And that some users will continue to use "start" on a non-existing machine

It is my understanding that this will continue to work as before (discouraged, but not deprecated) for backwards compatibility.

I'm not even sure why it needs to be "discouraged"; just not emphasizing it in the docs should be enough.

**Reply by jandubois** [reply]
> there would still be users that would like the "create" to also start it

I don't see the point. `limactl start` already does exactly this.

**Reply by jandubois** [reply]
If anything, I could see us adding a `--no-create` switch to `start` that will disable the already deprecated functionality that `limactl start foo` is equivalent to `limactl start --name foo template://default`.

I don't know if there is a plan to remove this deprecated functionality; I'm not sure it that is necessary and I generally dislike breaking backwards compatibility unless it is necessary to add new functionality.

Maybe we need to have a list of discouraged vs. deprecated functionality?

**Reply by jay7x** [reply]
My $2c:

1. It'd be nice to have `limactl create ... --start` which call the start action in the end.. (and `limactl stop --rm` too). But not really critical.
2. I think `limactl start` should continue to do the same as before but the options to create a VM should be hidden from the CLI help output. Maybe it'd be nice to throw a deprecation warning when it's used in this way too.

**Reply by jandubois** [reply]
> It'd be nice to have limactl create ... --start which call the start action in the end.. (and limactl stop --rm too). 

But why? This is completely redundant with `limactl start`.

Same thing for `limactl stop --rm`. This is the same as `limactl delete -f`, except slower because it would wait for a graceful shutdown before deleting the instance anyways. Why do we need an extra option for this?

Personally I don't like having many different subcommands and options that do the same thing[^1]. It is confusing for a person just learning about Lima and provides no actual value. I make exceptions for maintaining backwards compatibility, but am strongly against adding complexity for no reason beyond "because we can".

Subcommands/options should be orthogonal, composable, and easy to figure out.

[^1]: I realize that this is somewhat ironic, coming from a retired Perl maintainer.

**Reply by afbjorklund** [reply]
The only issue that I have is that I remove some instances to save spaces, and then forget that I have done so...

And then when it is not running, I try to start it and run into something similar to `limactl start docker`.

```
INFO[0000] Creating an instance "docker" from template://default (Not from template://docker) 
```

But at this point (backwards compatible), it might be hard to ever remove the create from the start anyway?

**Reply by afbjorklund** [reply]
The "others" do this, in case it was unclear _why_ I was expecting it:

```console
$ docker-machine start
Error: No machine name(s) specified and no "default" machine exists
```

```console
$ podman machine start
Error: podman-machine-default: VM does not exist
```

However, "minikube start" works the same way as "limactl start" does.

**Reply by AkihiroSuda** [reply, marked as answer]
PR:
- https://github.com/lima-vm/lima/pull/1677
